### PR TITLE
LoRA Builders for MM

### PR DESCRIPTION
**Pull request by pbontrager**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
- CLIP lora encoder + lora mlp + lora attention
- Flamingo lora encoder + decoder + projection head
- Llama 3.1 updated shared lora util
- TODO: update lora recipes to match full finetune updates

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/models/clip/_component_builders.py [hunk 1]
```diff
@@ -140,3 +152,334 @@ def clip_mlp(in_dim: int, out_dim: int, hidden_dim: int, activation: nn.Module,
...
+            )
+    else:
+        mlp = clip_mlp(
+            in_dim=embed_dim,
+            hidden_dim=4 * embed_dim,
+            out_dim=embed_dim,
+            activation=activation(),
+            quantize_base=quantize_base,
+        )
+    transformer_layer = TransformerSelfAttentionLayer(
```
Comment (felipemello1) [inline torchtune/models/clip/_component_builders.py:269]: do you mind using ModuleList like in llama 3.1, so we can avoid using _copy?

File: torchtune/models/flamingo/_component_builders.py [hunk 2]
```diff
@@ -197,21 +527,34 @@ def flamingo_decoder(
     for idx in range(1, num_layers + 1):
 
         # Self attention layers for text decoder
-        rope = RotaryPositionalEmbeddings(dim=head_dim, max_seq_len=max_seq_len, base=rope_base)
-        self_attn = MultiHeadAttention(
+        rope = Llama3ScaledRoPE(dim=head_dim, max_seq_len=max_seq_len, base=rope_base)
```
Comment (felipemello1) [inline torchtune/models/flamingo/_component_builders.py:530]: rope should be instantiated only once, ouside of the for loop, to avoid copies and extra memory

File: torchtune/models/clip/_component_builders.py (hunk 1 above)
Comment (pbontrager) [reply torchtune/models/clip/_component_builders.py:269]: I'd rather leave this for a future depreciation of the layer option

Comment (felipemello1) [reply torchtune/models/clip/_component_builders.py:269]: does it use rope embeddings? If so, it will consume more memory. Either way works for me

File: recipes/lora_finetune_single_device.py [hunk 3]
```diff
@@ -282,10 +284,12 @@ def setup(self, cfg: DictConfig) -> None:
 
         # Dataloader depends on the tokenizer and loss_fn and should be
         # setup after all of these are setup
+        collate_name = cfg.get("collate_fn", "torchtune.data.padded_collate_sft")
```
Comment (joecummings) [inline recipes/lora_finetune_single_device.py:287]: Why is padded_collate_sft the default? I thought that didn't work

File: torchtune/models/clip/_component_builders.py (hunk 1 above)
Comment (joecummings) [reply torchtune/models/clip/_component_builders.py:269]: Do it @pbontrager

File: recipes/lora_finetune_single_device.py (hunk 3 above)
Comment (pbontrager) [reply recipes/lora_finetune_single_device.py:287]: That's the standard collate for finetuning text models.

File: recipes/lora_finetune_distributed.py [hunk 4]
```diff
@@ -545,15 +555,20 @@ def _setup_data(
 
         if isinstance(cfg_dataset, ListConfig):
             datasets = [
-                config.instantiate(single_cfg_dataset, tokenizer=self._tokenizer)
+                config.instantiate(single_cfg_dataset, self._tokenizer)
```
Comment (ebsmothers) [inline recipes/lora_finetune_distributed.py:558]: This doesn't need to be keyword?

File: torchtune/models/clip/_component_builders.py [hunk 5]
```diff
@@ -140,3 +152,334 @@ def clip_mlp(in_dim: int, out_dim: int, hidden_dim: int, activation: nn.Module,
...
+    num_layers: int,
+    num_heads: int,
+    activation: Callable = nn.SiLU,
+    cls_output_dim: int = 512,
+    attn_bias: bool = True,
+    out_indices: Optional[List[int]] = None,
+    output_cls_projection: bool = False,
+    max_num_tiles: int = 4,
+    in_channels: int = 3,
+    intermediate_act: torch.nn.Module = torch.nn.SiLU(),
```
Comment (ebsmothers) [inline torchtune/models/clip/_component_builders.py:178]: Looking at L171 and L178 makes me sad

File: torchtune/models/clip/_component_builders.py [hunk 6]
```diff
@@ -140,3 +152,334 @@ def clip_mlp(in_dim: int, out_dim: int, hidden_dim: int, activation: nn.Module,
...
     return FeedForward(gate_proj=gate_proj, down_proj=down_proj, up_proj=None, activation=activation)
+
+
+# ------------------ LoRA CLIP ------------------
+
+
+def lora_clip_vision_encoder(
+    lora_modules: List[LORA_ATTN_MODULES],
+    apply_lora_to_mlp: bool = False,
+    apply_lora_to_output: bool = False,
```
Comment (ebsmothers) [inline torchtune/models/clip/_component_builders.py:163]: Do we just need to pass this for consistency even though it's a no-op? Would maybe raise an error if it's set to true or something

File: torchtune/models/flamingo/_encoder.py [hunk 7]
```diff
@@ -19,22 +18,20 @@ class FlamingoProjectionHead(nn.Module):
...
         self,
-        layer: nn.Module,
-        num_layers: int,
+        layers: nn.Module,
         output: nn.Module,
         num_hidden_inputs: int = 0,
     ) -> None:
         super().__init__()
-        self.layers = _get_clones(layer, num_layers)
+        self.layers = nn.ModuleList(layers)
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_encoder.py:34]: Shouldn't layers be `List[nn.Module]` type then?

File: torchtune/models/flamingo/_component_builders.py [hunk 8]
```diff
@@ -264,3 +238,507 @@ def flamingo_decoder(
...
+    return FlamingoProjectionHead(
+        layers=layers,
+        output=nn.Linear(proj_in, decoder_embed_dim),
+        num_hidden_inputs=num_hidden_inputs
+    )
+
+# ------------------ LoRA Flamingo ------------------
+
+
+class LoRATrainable(Enum):
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_component_builders.py:315]: Where is this used?

File: torchtune/models/flamingo/_component_builders.py [hunk 9]
```diff
@@ -264,3 +238,507 @@ def flamingo_decoder(
...
+    }
+
+    # clip encoder
+    clip_options = {
+        "tile_size": tile_size,
+        "patch_size": patch_size,
+        "embed_dim": clip_embed_dim,
+        "num_layers": clip_num_layers,
+        "num_heads": num_heads,
+        "activation": nn.GELU,
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_component_builders.py:414]: Maybe it's deliberate but seems like this doesn't match the default in the CLIP builder

File: torchtune/models/clip/_component_builders.py (hunk 6 above)
Comment (felipemello1) [reply torchtune/models/clip/_component_builders.py:163]: i like consistency, agree with raising an error saying that it is a no-op/not supported. We do it for all tied embedding models that dont support apply_lora_to_output

File: torchtune/models/flamingo/_component_builders.py [hunk 10]
```diff
@@ -264,3 +238,507 @@ def flamingo_decoder(
...
+        projection_head = lora_flamingo_projection_head(**projection_options, **lora_options)
+    else:
+        projection_head = flamingo_projection_head(**projection_options)
+
+    return FlamingoEncoder(clip=clip, projection_head=projection_head)
+
+
+def lora_flamingo_decoder(
+    decoder_lora: bool,
+    fusion_lora: bool,
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_component_builders.py:444]: not used?

File: torchtune/models/flamingo/_encoder.py (hunk 7 above)
Comment (felipemello1) [reply torchtune/models/flamingo/_encoder.py:34]: please look at the transformer, that supports all cases (List[nn.Module], nn.ModuleList, nn.module: https://github.com/pytorch/torchtune/blob/30b8519e86d2094879576e16ab1f8d21a6f6f9f7/torchtune/modules/transformer.py#L353

I am fine if in flamingo it only supports nn.ModuleList though, but i prefer the consistency

File: torchtune/models/flamingo/_component_builders.py [hunk 11]
```diff
@@ -264,3 +238,507 @@ def flamingo_decoder(
...
+                lora_dropout=lora_dropout,
+                use_dora=use_dora,
+            )
+        else:
+            mlp = llama3_mlp(dim=embed_dim, hidden_dim=hidden_dim, quantize_base=quantize_base)
+        decoder_layer = TransformerSelfAttentionLayer(
+            attn=self_attn,
+            mlp=mlp,
+            sa_norm=RMSNorm(dim=embed_dim, eps=1e-5),
+            mlp_norm=RMSNorm(dim=embed_dim, eps=1e-5),
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_component_builders.py:554]: super nit: use the same format for eps (below it's `1e-05`)

File: torchtune/models/flamingo/_component_builders.py [hunk 12]
```diff
@@ -264,3 +238,507 @@ def flamingo_decoder(
...
+    *,
+    # projection head parameters
+    num_layers: int,
+    num_heads: int,
+    decoder_embed_dim: int,
+    clip_embed_dim: int,
+    num_hidden_inputs: int,
+    # LoRA args
+    apply_lora_to_mlp: bool,
+    apply_lora_to_output: bool,
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_component_builders.py:644]: Any particular reason these are now keyword-only args as opposed to how we have them elsewhere?

File: torchtune/models/flamingo/_component_builders.py [hunk 13]
```diff
@@ -264,3 +238,507 @@ def flamingo_decoder(
...
+            attn=self_attn,
+            mlp=mlp,
+            sa_norm=RMSNorm(dim=embed_dim, eps=1e-5),
+            mlp_norm=RMSNorm(dim=embed_dim, eps=1e-5),
+        )
+
+        # cross attention layers, mixing text and vision,
+        # placed every `fusion_interval` layers
+        if idx % fusion_interval == 0:
+            attn = lora_llama3_attention(
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_component_builders.py:560]: Could maybe use partials to reduce the duplicative code here? But nbd either way

File: torchtune/models/flamingo/_component_builders.py [hunk 14]
```diff
@@ -264,3 +238,507 @@ def flamingo_decoder(
...
+                    lora_dropout=lora_dropout,
+                    use_dora=use_dora,
+                )
+            else:
+                mlp = llama3_mlp(dim=embed_dim, hidden_dim=hidden_dim, quantize_base=quantize_base)
+            xattn_layer = TransformerCrossAttentionLayer(
+                attn=attn,
+                mlp=mlp,
+                ca_norm=RMSNorm(dim=embed_dim),
+                mlp_norm=RMSNorm(dim=embed_dim),
```
Comment (ebsmothers) [inline torchtune/models/flamingo/_component_builders.py:594]: Is it deliberate to have different eps here vs in self-attention layers?

**Review by ebsmothers — APPROVED** [review]

File: recipes/lora_finetune_distributed.py (hunk 4 above)
Comment (pbontrager) [reply recipes/lora_finetune_distributed.py:558]: Some datasets call it tokenizer and others call it transforms

File: torchtune/models/clip/_component_builders.py (hunk 6 above)
Comment (pbontrager) [reply torchtune/models/clip/_component_builders.py:163]: What can't they be true?

File: torchtune/models/flamingo/_component_builders.py (hunk 9 above)
Comment (pbontrager) [reply torchtune/models/flamingo/_component_builders.py:414]: Deliberate

File: torchtune/models/flamingo/_component_builders.py (hunk 10 above)
Comment (pbontrager) [reply torchtune/models/flamingo/_component_builders.py:444]: will be used in builders
